static/js/drafts.js

```diff
@@ -432,8 +482,82 @@ export function launch() {
         });
     }
 
-    const drafts = format_drafts(draft_model.get());
-    render_widgets(drafts);
+    function filter_drafts_by_compose_box_and_recipient(drafts) {
+        // Prioritize stream, topic, and PM recipients from the compose box first and then
+        // from the narrow.
+        const composing = compose_state.composing();
+        let stream_id = null;
+        let topic = null;
+        if (composing) {
+            if (compose_state.get_message_type() === "stream") {
+                const stream_name = compose_state.stream_name();
+                stream_id = stream_name ? stream_data.get_stream_id(stream_name) : null;
+                topic = compose_state.topic();
+            }
+        } else {
+            const stream_name = narrow_state.stream();
+            stream_id = stream_name ? stream_data.get_stream_id(stream_name) : null;
+            topic = narrow_state.topic();
+        }
+
+        let private_recipients = null;
+        if (composing) {
+            if (compose_state.get_message_type() === "private") {
+                private_recipients = compose_state.private_message_recipient();
+            }
+        } else {
+            private_recipients = narrow_state.pm_emails_string();
+        }
```

Comment:
What's the thinking around the structure of the 4 cases you have here? Maybe this is the cleanest way to do it, but it feels hard to read with the first if/else being very parallel to the second.

It's possible this would read better as a function returning an object, since then you can return things in order to avoid the if/else chaining getting crazy.

Reply:
Yeah there's a lot of mixing and matching of cases and I kept switching around how I ordered them not really liking any of them. I like the idea of moving that stuff to its own function :)